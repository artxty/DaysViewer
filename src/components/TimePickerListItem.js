import PropTypes from 'prop-types';
import React, { useState } from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
} from 'react-native';
import DateTimePicker from '@react-native-community/datetimepicker';
import iosColors from 'ios-colors';

import { format } from '../helpers/date';

const styles = StyleSheet.create({
  item: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  text: {
    fontSize: 18,
  },
  timeText: {
    color: iosColors.systemBlue.light,
  },
});

export default function TimePickerItem({ value, onChange, style }) {
  const [isTimePickerVisible, setIsTimePickerVisible] = useState(false);
  const [time, setTime] = useState(new Date(0, 0, 0, value.hours, value.minutes));

  function handleChange(event, selectedTime) {
    const currentTime = selectedTime || time;
    setTime(currentTime);
    onChange({
      hours: currentTime.getHours(),
      minutes: currentTime.getMinutes(),
    });
  }

  return (
    <View style={style}>
      <TouchableOpacity onPress={() => setIsTimePickerVisible(!isTimePickerVisible)}>
        <View style={styles.item}>
          <Text style={styles.text}>Time</Text>
          <Text style={{ ...styles.text, ...styles.timeText }}>{format(time, 'p')}</Text>
        </View>
      </TouchableOpacity>
      {isTimePickerVisible && (
        <DateTimePicker
          testID="dateTimePicker"
          value={time}
          mode="time"
          display="default"
          onChange={handleChange}
        />
      )}
    </View>
  );
}

TimePickerItem.propTypes = {
  value: PropTypes.shape({
    hours: PropTypes.number,
    minutes: PropTypes.number,
  }).isRequired,
  onChange: PropTypes.func.isRequired,
  style: PropTypes.shape({}),
};

TimePickerItem.defaultProps = {
  style: StyleSheet.create({}),
};
