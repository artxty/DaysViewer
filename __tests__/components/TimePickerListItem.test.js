import React from 'react';
import renderer, { act } from 'react-test-renderer';

import TimePickerListItem from '../../src/components/TimePickerListItem';

jest.mock('@react-native-community/datetimepicker', () => 'DateTimePicker');

jest.mock('react-native', () => {
  const reactNative = jest.requireActual('react-native');
  return {
    __esModule: true,
    TouchableOpacity: 'TouchableOpacity',
    View: 'View',
    Text: 'Text',
    StyleSheet: reactNative.StyleSheet,
  };
});

describe('TimePickerListItem', () => {
  let tree;
  let touchableOpacity;
  const onChangeMock = jest.fn();
  beforeEach(() => {
    tree = renderer.create(<TimePickerListItem
      value={{ hours: 4, minutes: 20 }}
      onChange={onChangeMock}
    />);
    touchableOpacity = tree.root.find((el) => el.type === 'TouchableOpacity');
    jest.clearAllMocks();
  });

  it('renders correctly when DateTimePicker is hidden', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should appear DateTimePicker after onPress', () => {
    act(() => {
      touchableOpacity.props.onPress();
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  describe('DateTimePicker', () => {
    let dateTimePicker;
    beforeEach(() => {
      act(() => {
        touchableOpacity.props.onPress();
      });
      dateTimePicker = tree.root.find((el) => el.type === 'DateTimePicker');
    });

    it('should call props.onChange when own onChange called', () => {
      act(() => {
        dateTimePicker.props.onChange();
      });

      expect(onChangeMock).toBeCalledTimes(1);
    });

    it('should call props.onChange with correct arguments', () => {
      /* should transform Date object to { hours, minutes } object */
      const expectedArgs = { hours: 8, minutes: 40 };
      const fakeDateAfterChange = new Date(0, 0, 0, expectedArgs.hours, expectedArgs.minutes);

      act(() => {
        dateTimePicker.props.onChange(null, fakeDateAfterChange);
      });

      expect(onChangeMock).toBeCalledWith(expectedArgs);
    });
  });
});
