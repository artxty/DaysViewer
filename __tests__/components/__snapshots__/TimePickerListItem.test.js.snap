// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePickerListItem renders correctly when DateTimePicker is hidden 1`] = `
<View
  style={Object {}}
>
  <TouchableOpacity
    onPress={[Function]}
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        Time
      </Text>
      <Text
        style={
          Object {
            "color": "rgb(0, 122, 255)",
            "fontSize": 18,
          }
        }
      >
        4:20 AM
      </Text>
    </View>
  </TouchableOpacity>
</View>
`;

exports[`TimePickerListItem should appear DateTimePicker after onPress 1`] = `
<View
  style={Object {}}
>
  <TouchableOpacity
    onPress={[Function]}
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        Time
      </Text>
      <Text
        style={
          Object {
            "color": "rgb(0, 122, 255)",
            "fontSize": 18,
          }
        }
      >
        4:20 AM
      </Text>
    </View>
  </TouchableOpacity>
  <DateTimePicker
    display="default"
    mode="time"
    onChange={[Function]}
    testID="dateTimePicker"
    value={1899-12-31T04:20:00.000Z}
  />
</View>
`;
